lib/remote_io.rb

```diff
@@ -73,63 +79,89 @@ def read(n_bytes)
 
   protected
 
+  REDIRECT_LIMIT = 3
+  UNSAFE_URI_CHARS = %r{[^\-_.!~*'()a-zA-Z\d;/?:@&=+$,\[\]%]}
+
+  # Remove the fragment and escape any unsafe characters in the given URI,
+  # retaining the prior host and scheme if the URI is relative.
+  #
+  # @param uri[String, URI::Generic] The URI to be made safe.
+  def uri=(uri)
```

Comment:
a. The intention here is to replicate what Faraday does under the hood when handling redirection. We escape any unsafe characters, and handles any cases where relative paths are returned.

b. We want to do this every time we get redirected - meaning every time we update the `uri` attribute; so it felt appropriate to implement this using a setter so that we can feel confident the URI in use will be valid and pointing to the correct location. It future-proofs it a bit in-case we end up setting the `uri` attribute elsewhere. Feels clean to me, but I'm okay with inlining it or placing it in a simple standalone method if you disagree.

Reply:
Got it 👌